types/react-redux/index.d.ts

```diff
@@ -141,7 +142,7 @@ interface MapStateToPropsFactory<TStateProps, TOwnProps> {
     (initialState: any, ownProps: TOwnProps): MapStateToProps<TStateProps, TOwnProps>;
 }
 
-type MapStateToPropsParam<TStateProps, TOwnProps> = MapStateToProps<TStateProps, TOwnProps> | MapStateToPropsFactory<TStateProps, TOwnProps>;
+type MapStateToPropsParam<TStateProps, TOwnProps> = MapStateToPropsFactory<TStateProps, TOwnProps> | MapStateToProps<TStateProps, TOwnProps> | null | undefined;
```

Comment:
The previous order caused a map factory to not be inferred. Listing the factory first forces TS to recognized the factory first and thus infer `TStateProps` properly

Reply:
late to the 🎉 :
was adding `| null | undefined` necessary? any reason you didn't do that for `MapDispatchToPropsParam`?

Reply:
It's necessary because often times you only want to give an argument for dispatch without requiring state which means you need to pass null for `mapStateToProps`.

Reply:
wouldn't the same be true for only providing an argument for `mergeProps`?

```
export const FooHOC = connect<undefined, undefined, TOwnProps, TMergedProps>(
    undefined,
    undefined,
    mergeProps,
)(UnconnectedFoo);
```

I have similar code in a codebase with `strictNullChecks`, and it works fine with current typings.

Reply:
on second thought, this goes against the overloads we've provided.
the idea is that overloads that expect `mapStateToProps` to be defined reference this type, and those who don't expect `null | undefined` explicitly.

e.g.
https://github.com/NicholasBoll/DefinitelyTyped/blob/1cc622c196e2451bd2272e2af87221fc82da2dec/types/react-redux/index.d.ts#L76
vs
https://github.com/NicholasBoll/DefinitelyTyped/blob/1cc622c196e2451bd2272e2af87221fc82da2dec/types/react-redux/index.d.ts#L81

Reply:
Good catch. I don't think the changes are going to cause any issues but we should probably change this one piece back to the way that it was since it's unnecessary.

Reply:
Good catch. Originally I was able to get right of 4 extra overloads, but decided against the more drastic change. The `null | undefined` was leftover from that revert. It doesn't actually cause harm in this case, but currently the overloads don't work as expected anyways. Oftentimes the unexpected overload is picked (`connect(mapStateToProps, mapStateToDispatch)(Component)` will pick an overload with `mergeProps`. I was also experimenting forcing some overloads to extend from `{}` or `object` to force Typescript to recognize the correct overload.

I thought we should add more test cases before refactoring anything though. It should be safe to remove `null | undefined` as that edge case shouldn't be hit with the current typings.